Application powerwall for parameter graphing

CWNE88 opened the pull request:
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/librenms/librenms?pullRequest=8526) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/librenms/librenms?pullRequest=8526) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/librenms/librenms?pullRequest=8526) it.</sub>

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/9e52b906-7efb-4d6c-a450-c1cca4517f95) completed: **5 new issues**

murrant commented on html/includes/graphs/application/powerwall_battery.inc.php:
```diff
@@ -0,0 +1,24 @@
+
+$scale_min = '0';
+
+$rrd_filename = rrd_name($device['hostname'], 'powerwall_battery');
```
This is the wrong rrd name.  your rrds are named `app-powerwall-3.rrd` for example. Look at the other application based graphs to see how.

also, `var_dump($variable);` to see what a variable is set too. much better than guessing.

laf replied:
No murrant is saying that your poller file is creating the file name like `$rrd_name = array('app', $name, $app_id);` which results in a file called `app-powerwall-3.rrd` so you need to build the same filename in the graph code as well otherwise the graphs won't work / show anything.

In general the `var_dump()` is a function you can use in testing to print out the variables for you to see what they are called so `var_dump($rrd_filename);` will print out what the filename is.

CWNE88 replied:
Are you sure my poller application is even right to begin with? Most of it was guesswork

laf replied:
Looks fine. Some optimisations need to be done but get it working first.

murrant replied:
This is literally what you need:
`$rrd_filename = rrd_name($device['hostname'], array('app', $app['app_type'], $app['app_id']));`

You can find that by looking at most of the other application graphs.

CWNE88 replied:
I managed to get it working last night
https://snag.gy/li0gZE.jpg
https://snag.gy/KaWc57.jpg

I'll add the other parameters today and hopefully tidy it all up.

CWNE88 replied:
I now have a basic system working, but feel there are too many graphs, so I want to cull some off first.
There are currently 17 graph files, which is excessive, and there are a couple that I do want to add, but they are COUNTER not GAUGE, so I'm not sure how to represent them (energy usage, so can hopefully see daily usage)

Also, trying to work out the multiple line graph, to make a 4 in 1 overview graph.

murrant commented:
To make this discoverable, you should use one extend and return all data.

CWNE88 commented:
I'm currently putting the variables into an array with addDataset and getting that to work, then I should have something to show.

murrant commented:
@CWNE88 did you ever finish this?  Please give us an update.

CWNE88 commented:
Yes, I uploaded my files using the web github to their locations (or at least I thought I did!)
The result looks like this:
https://snag.gy/mKz0tE.jpg
https://snag.gy/j2ZALO.jpg
https://snag.gy/YgihTo.jpg
https://snag.gy/UKV1Zm.jpg
https://snag.gy/Xs6cG5.jpg

The only thing I never got working, was those imported/exported values I think should have been counter, not gauge, as they will only ever go up. There was always 'U' in the RRD data for them though, so it didn't seem to work.
By using gauge though, I can still see how much energy has been imported/exported within a time frame by looking at difference between min and max. I don't know how counter presents itself anyway.

All values for power ever passed via SNMP are rounded to the nearest 50 before sending, so I guess the averages over time cause the odd numbers, which is fine. The summary graph doesn't need decimal places though, but it has 2 for some reason.

murrant commented:
You would need to add them to your powerwall branch instead of your master branch, here: https://github.com/CWNE88/librenms/tree/powerwall

Do you intend to finish?

CWNE88 commented:
I'll add them there tomorrow.

Since I couldn't figure out the rrd counter issue, I've gone about as far as I can to make it operational and usable as is.

murrant commented:
I can take a look once you push your updated files.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/cdcbd897-45fe-4635-b973-1ab4de1f0d95) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/cdcbd897-45fe-4635-b973-1ab4de1f0d95) completed: **6 new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5df9db50-5f43-46c7-a372-3d7a9826b91c) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/1537505b-383b-4b13-9862-87e537a7ddc2) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/93793774-7090-4f7e-a052-da56508186c4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/93793774-7090-4f7e-a052-da56508186c4) completed: **1 new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/13f54258-bb0d-4853-8813-d55bd42ed64e) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/2b17feba-5209-405d-bd46-9a65cee9ae16) was created.

laf commented:
I've updated this to fix the travis issues.

Have you done the librenms-agent code as well?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/2b17feba-5209-405d-bd46-9a65cee9ae16) completed: **1 new issues**

CWNE88 commented:
![solar rounding](https://user-images.githubusercontent.com/38122076/39817741-54dd61ba-53e2-11e8-962c-a162918c0aac.jpg)

One thing I've just noticed is rounding in the various energy imported/exported once it got to 1000k. I want it to keep showing k, for instance this current/max value should show 1017 (from value of 1016899). That way, I can see that 25kWh got exported today. Once it rounds it, I can't tell

murrant commented:
That is in the base graph you are including, there might be a variable you can set or something to avoid that.

CWNE88 commented:
Do you mean where this line points to:
require 'includes/graphs/generic_v3_multiline.inc.php';

...so I checked that file too but don't see anything about rounding values

laf commented:
> ...so I checked that file too but don't see anything about rounding values

You might want:

`[-X|--units-exponent value]`

https://oss.oetiker.ch/rrdtool/doc/rrdgraph.en.html

We don't currently use it but try passing it to `$rrd_options`



> On a Linux box, I use the extend powerwall in snmpd to return the data.

Then you need to include this script in the librenms/librenms-agent repo. This will also need the https://docs.librenms.org/#Extensions/Applications/ updating to include this as well.

CWNE88 commented:
The looks exactly like what I'm after according to the docs  [-X|--units-exponent value]
..but where would that be used? I assume the file powerwall_solar-exported.inc.php but I can't see the command rrdtool itself in that file, so can't figure out where this syntax is actually used.

I'll put the powerwall script that returns the data as well.

laf commented:
> The looks exactly like what I'm after according to the docs [-X|--units-exponent value]
> ..but where would that be used? I assume the file powerwall_solar-exported.inc.php but I can't see the command rrdtool itself in that file, so can't figure out where this syntax is actually used.

You can possibly just add it to the end of the file with `$rrd_options .= ' -X blah';`

CWNE88 commented:
I tried adding it to "powerwall_solar-exported.inc.php"
$rrd_options = '-X 10e3';

But it did nothing.

I also tried with .=  exactly as per your message too, but it had error drawing graphs at all.

![rounding](https://user-images.githubusercontent.com/38122076/40029674-2660fa22-5829-11e8-87cd-abc70f052c21.jpg)

CWNE88 commented:
I tried:
$rrd_options .= '-X 10e3';

and the rrd output is:
RRDTool Command

rrdtool graph /tmp/o8bDdq0fERQJpsRx -X 10e3 --alt-autoscale-max --rigid -E --start 1526349900 --end 1526436300 --width 1152 --height 300 -c BACK#EEEEEE00 -c SHADEA#EEEEEE00 -c SHADEB#EEEEEE00 -c FONT#000000 -c CANVAS#FFFFFF00 -c GRID#a5a5a5 -c MGRID#FF9999 -c FRAME#5e5e5e -c ARROW#5e5e5e -R normal --font LEGEND:8:DejaVuSansMono --font AXIS:7:DejaVuSansMono --font-render-mode normal COMMENT:'Solar Exported              Now         Min         Max        Avg\l' COMMENT:'\l' DEF:solar-exported1=/opt/librenms/rrd/localhost/app-powerwall-68.rrd:solar-exported:AVERAGE  DEF:solar-exported1min=/opt/librenms/rrd/localhost/app-powerwall-68.rrd:solar-exported:MIN  DEF:solar-exported1max=/opt/librenms/rrd/localhost/app-powerwall-68.rrd:solar-exported:MAX  LINE2:solar-exported1#ffcc00:'Watt Hours         ' AREA:solar-exported1#ffcc0033:'' GPRINT:solar-exported1:LAST:%8.0lf%s GPRINT:solar-exported1min:MIN:%8.0lf%s GPRINT:solar-exported1max:MAX:%8.0lf%s GPRINT:solar-exported1:AVERAGE:'%8.0lf%s\n' COMMENT:'\n'

RRDTool Output

1233x369
OK u:0.06 s:0.01 r:0.06

Values still all show 1M though, not the k values

murrant commented:
Docs say the value should be an integer, so just ` -X 3` I think.

CWNE88 commented:
As per the docs, it seems I just need the value 3 for kilo.
So I put this at the end of powerwall_solar-exported.inc.php
$rrd_options .= '-X 3';

The graph didn't work, and said:

RRDTool Command

rrdtool graph /tmp/ddYVkypf1T44gG6r  --alt-autoscale-max --rigid -E --start 1526375100 --end 1526461500 --width 1912.5 --height 563 -c BACK#EEEEEE00 -c SHADEA#EEEEEE00 -c SHADEB#EEEEEE00 -c FONT#000000 -c CANVAS#FFFFFF00 -c GRID#a5a5a5 -c MGRID#FF9999 -c FRAME#5e5e5e -c ARROW#5e5e5e -R normal --font LEGEND:8:DejaVuSansMono --font AXIS:7:DejaVuSansMono --font-render-mode normal --alt-autoscale-max --rigid -E --start 1526375100 --end 1526461500 --width 1912.5 --height 563 -c BACK#EEEEEE00 -c SHADEA#EEEEEE00 -c SHADEB#EEEEEE00 -c FONT#000000 -c CANVAS#FFFFFF00 -c GRID#a5a5a5 -c MGRID#FF9999 -c FRAME#5e5e5e -c ARROW#5e5e5e -R normal --font LEGEND:8:DejaVuSansMono --font AXIS:7:DejaVuSansMono --font-render-mode normal COMMENT:'Solar Exported              Now         Min         Max        Avg\l' COMMENT:'\l' DEF:solar-exported1=/opt/librenms/rrd/localhost/app-powerwall-68.rrd:solar-exported:AVERAGE  DEF:solar-exported1min=/opt/librenms/rrd/localhost/app-powerwall-68.rrd:solar-exported:MIN  DEF:solar-exported1max=/opt/librenms/rrd/localhost/app-powerwall-68.rrd:solar-exported:MAX  LINE2:solar-exported1#ffcc00:'Watt Hours         ' AREA:solar-exported1#ffcc0033:'' GPRINT:solar-exported1:LAST:%8.0lf%s GPRINT:solar-exported1min:MIN:%8.0lf%s GPRINT:solar-exported1max:MAX:%8.0lf%s GPRINT:solar-exported1:AVERAGE:'%8.0lf%s\n' COMMENT:'\n'-X 3

RRDTool Output

ERROR: Could not make sense out of '3'

I see in the command it says COMMENT... should it really be a comment?

CWNE88 commented:
This was actually originally supposed to be a counter not gauge, but I'll stick with gauge if I can get it in kilo

laf commented:
You need a space before the `-X` like I showed.

CWNE88 commented:
I've tried using COUNTER for some of the RRD values, as that's what they should be.

If I use counter though, nothing at all gets shown from the RRD file, even the gauges that used to work.

This is an example with all values as gauge:
https://pastebin.com/6nJqkv2P

And this example is with some as counter:
https://pastebin.com/6XxgLYF5

Any ideas why this is happening?

laf commented:
Honestly you should push your code up so we can see what it's doing.

CWNE88 commented:
I uploaded it all a few weeks ago.
You updated something about them it seemed too, so I don't know what you mean.


@laf
Update powerwall_summary.inc.php
a6a1167
@laf
Update powerwall.inc.php
90823d8 



I'll upload this new COUNTER version as well to show it

CWNE88 commented:
I just tried to upload, but I only see librenms/librenms, not the librenms/cwne88 fork that I had before so I couldn't upload it

CWNE88 commented:
The problem with the rrd file was that the counter value I was putting in had decimal places. Rounding it to an integer now works.

So now, how can I use that counter info? If I graph it I get strange numbers. By default is it graphing rate or change or something?

What I'd really like is to display in the app the value that it currently is, and how much it's increased from say the start of the day, or start of week (whatever graph period is displayed). Is that possible?

laf commented:
> What I'd really like is to display in the app the value that it currently is, and how much it's increased from say the start of the day, or start of week (whatever graph period is displayed). Is that possible?

You can't do that (maybe at all). You'd need to check wether rrdtool supported it before seeing if you can get it to work in LibreNMS.

laf commented:
Closing this down for now, no feedback in over a month. Feel free to respond and we can re-open.

CWNE88 commented:
I've just deleted the old files, and updated the new complete files (and only a few this time).

I also have the daemon which runs on a box to grab the actual data, and a readme overview (but wouldn't exactly call them instructions at this point).

Where should I upload the daemon and nanny scripts?

I'll write up an instruction summary of the application now.

laf commented:
The agent scripts go in this repo: https://github.com/librenms/librenms-agent/

CWNE88 commented:
I tried to upload (via web), but it says uploads are disabled.
Do I have to do a pull request or something for this repo?

CWNE88 commented:
This may take some time to figure out
